src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,232 @@
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm [comm_semigroup α] [add_comm_monoid α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant w ⬝ circulant v
+| 0     := dec_trivial
+| (n+1) := circulant_mul_comm
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub I] [has_scalar R α] {k : R} {v : I → α} :
+  circulant (k • v) = k • circulant v :=
+by {ext, simp [circulant]}
+
+@[simp] lemma circulant_zero [has_zero α] [has_sub I]:
+  circulant 0 = (0 : matrix I I α) :=
+by ext; simp [circulant]
+
+@[simp] lemma circulant_single [has_zero α] [has_one α] [decidable_eq I] [add_group I]:
+  circulant (pi.single 0 1 : I → α)= (1 : matrix I I α) :=
+by { ext i j, simp only [circulant, one_apply, pi.single_apply, sub_eq_zero], congr }
+
+/-- The identity matrix is a circulant matrix. -/
+lemma one_eq_circulant [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  (1 : matrix I I α) = circulant (λ i, ite (i = 0) 1 0) :=
+by { rw [←circulant_single], congr, ext i, simp [pi.single, update] }
+
+/-- Another version of `one_eq_circulant`. -/
+lemma one_eq_circulant' [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  (1 : matrix I I α) = circulant (λ i, (1 : matrix I I α) i 0) :=
+by { ext i j, simp only [circulant, one_apply, sub_eq_zero], congr }
+
+lemma fin.one_eq_circulant [has_zero α] [has_one α] :
+  ∀ n, (1 : matrix (fin n) (fin n) α) = circulant (λ i, ite (↑i = 0) 1 0)
+| 0     := dec_trivial
+| (n+1) :=
+begin
+  ext i j,
+  simp only [circulant, one_apply, ← @fin.coe_zero n, fin.coe_injective.eq_iff, sub_eq_zero],
+  congr,
+end
+
+/-- For a one-ary predicate `p`, `p` applied to every entry of `circulant v` is true
+    if `p` applied to every entry of `v` is true. -/
+lemma pred_circulant_entry_of_pred_vec_entry [has_sub I] {p : α → Prop} {v : I → α} :
+  (∀ k, p (v k)) → ∀ i j, p ((circulant v) i j) :=
+begin
+  intros h i j,
+  simp [circulant],
+  exact h (i - j),
+end
+
+/-- Given a set `S`, every entry of `circulant v` is in `S` if every entry of `v` is in `S`. -/
+lemma circulant_entry_in_of_vec_entry_in [has_sub I] {S : set α} {v : I → α} :
+  (∀ k, v k ∈ S) → ∀ i j, (circulant v) i j ∈ S :=
+@pred_circulant_entry_of_pred_vec_entry α I _ S v
+
+/-- The circulant matrix `circulant v` is symmetric iff `∀ i j, v (j - i) = v (i - j)`. -/
+lemma circulant_is_symm_ext_iff' [has_sub I] {v : I → α} :
+  (circulant v).is_symm ↔ ∀ i j, v (j - i) = v (i - j) :=
+by simp [is_symm.ext_iff, circulant]
+
+/-- The circulant matrix `circulant v` is symmetric iff `v (- i) = v i` if `[add_group I]`. -/
+lemma circulant_is_symm_ext_iff [add_group I] {v : I → α} :
+  (circulant v).is_symm ↔ ∀ i, v (- i) = v i :=
+begin
+  rw [circulant_is_symm_ext_iff'],
+  split,
+  { intros h i, convert h i 0; simp },
+  { intros h i j, convert h (i - j), simp }
+end
+
+lemma fin.circulant_is_sym_ext_iff :
```

Comment:
All the names of these lemmas talk about `is_sym`, but the condition is now called `is_symm`.